File: packages/sdkwork-birdcoder-codeengine/src-host/src/catalog.rs
use std::sync::OnceLock;

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum NativeSessionDiscoveryModeRecord {
    ExplicitOnly,
    PassiveGlobal,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodeEngineCapabilityMatrixRecord {
    pub chat: bool,
    pub streaming: bool,
    pub structured_output: bool,
    pub tool_calls: bool,
    pub planning: bool,
    pub patch_artifacts: bool,
    pub command_artifacts: bool,
    pub todo_artifacts: bool,
    pub pty_artifacts: bool,
    pub preview_artifacts: bool,
    pub test_artifacts: bool,
    pub approval_checkpoints: bool,
    pub session_resume: bool,
    pub remote_bridge: bool,
    pub mcp: bool,
}

#[derive(Clone, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PartialCodeEngineCapabilityMatrixRecord {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub chat: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub streaming: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub structured_output: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub planning: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub patch_artifacts: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub command_artifacts: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub todo_artifacts: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pty_artifacts: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub preview_artifacts: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub test_artifacts: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval_checkpoints: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_resume: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub remote_bridge: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mcp: Option<bool>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodeEngineDescriptorRecord {
    pub engine_key: String,
    pub display_name: String,
    pub vendor: String,
    pub installation_kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_model_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub homepage: Option<String>,
    pub supported_host_modes: Vec<String>,
    pub transport_kinds: Vec<String>,
    pub capability_matrix: CodeEngineCapabilityMatrixRecord,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub access_plan: Option<CodeEngineAccessPlanRecord>,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum CodeEngineAccessStrategyKindRecord {
    RustNative,
    GrpcBridge,
    OpenapiProxy,
    RemoteControl,
    CliSpawn,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum CodeEngineAccessLaneStatusRecord {
    Ready,
    Planned,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum CodeEngineRuntimeOwnerRecord {
    RustServer,
    TypescriptBridge,
    ExternalService,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum CodeEngineBridgeProtocolRecord {
    Direct,
    Grpc,
    Http,
    Websocket,
    Stdio,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodeEngineAccessLaneRecord {
    pub lane_id: String,
    pub label: String,
    pub strategy_kind: CodeEngineAccessStrategyKindRecord,
    pub runtime_owner: CodeEngineRuntimeOwnerRecord,
    pub bridge_protocol: CodeEngineBridgeProtocolRecord,
    pub transport_kind: String,
    pub status: CodeEngineAccessLaneStatusRecord,
    pub enabled_by_default: bool,
    pub host_modes: Vec<String>,
    pub description: String,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodeEngineAccessPlanRecord {
    pub primary_lane_id: String,
    pub fallback_lane_ids: Vec<String>,
    pub lanes: Vec<CodeEngineAccessLaneRecord>,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodeEngineModelCatalogEntryRecord {
    pub engine_key: String,
    pub model_id: String,
    pub display_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider_id: Option<String>,
    pub status: String,
    pub default_for_engine: bool,
    pub transport_kinds: Vec<String>,
    pub capability_matrix: PartialCodeEngineCapabilityMatrixRecord,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeSessionProviderCatalogRecord {
    pub engine_id: String,
    pub display_name: String,
    pub native_session_id_prefix: String,
    pub transport_kinds: Vec<String>,
    pub discovery_mode: NativeSessionDiscoveryModeRecord,
}

#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SharedCodeEngineCatalogRecord {
    pub engines: Vec<CodeEngineDescriptorRecord>,
    pub models: Vec<CodeEngineModelCatalogEntryRecord>,
    #[serde(default)]
    pub native_providers: Vec<NativeSessionProviderCatalogRecord>,
}

fn fallback_native_session_provider_catalog_entries(
) -> &'static [NativeSessionProviderCatalogRecord] {
    static FALLBACK_NATIVE_SESSION_PROVIDER_CATALOG: OnceLock<
        Vec<NativeSessionProviderCatalogRecord>,
    > = OnceLock::new();
    FALLBACK_NATIVE_SESSION_PROVIDER_CATALOG
        .get_or_init(|| {
            vec![
                NativeSessionProviderCatalogRecord {
                    engine_id: "codex".to_owned(),
                    display_name: "Codex".to_owned(),
                    native_session_id_prefix: "codex-native:".to_owned(),
                    transport_kinds: vec!["sdk-stream".to_owned(), "cli-jsonl".to_owned()],
                    discovery_mode: NativeSessionDiscoveryModeRecord::PassiveGlobal,
                },
                NativeSessionProviderCatalogRecord {
                    engine_id: "claude-code".to_owned(),
                    display_name: "Claude Code".to_owned(),
                    native_session_id_prefix: "claude-code-native:".to_owned(),
                    transport_kinds: vec![
                        "sdk-stream".to_owned(),
                        "remote-control-http".to_owned(),
                    ],
                    discovery_mode: NativeSessionDiscoveryModeRecord::ExplicitOnly,
                },
                NativeSessionProviderCatalogRecord {
                    engine_id: "gemini".to_owned(),
                    display_name: "Gemini".to_owned(),
                    native_session_id_prefix: "gemini-native:".to_owned(),
                    transport_kinds: vec!["sdk-stream".to_owned(), "openapi-http".to_owned()],
                    discovery_mode: NativeSessionDiscoveryModeRecord::ExplicitOnly,
                },
                NativeSessionProviderCatalogRecord {
                    engine_id: "opencode".to_owned(),
                    display_name: "OpenCode".to_owned(),
                    native_session_id_prefix: "opencode-native:".to_owned(),
                    transport_kinds: vec![
                        "sdk-stream".to_owned(),
                        "openapi-http".to_owned(),
                        "cli-jsonl".to_owned(),
                    ],
                    discovery_mode: NativeSessionDiscoveryModeRecord::ExplicitOnly,
                },
            ]
        })
        .as_slice()
}

pub fn native_session_provider_catalog_entries(
) -> &'static [NativeSessionProviderCatalogRecord] {
    let generated_entries = shared_codeengine_catalog().native_providers.as_slice();
    if generated_entries.is_empty() {
        return fallback_native_session_provider_catalog_entries();
    }

    generated_entries
}

pub fn shared_codeengine_catalog() -> &'static SharedCodeEngineCatalogRecord {
    static SHARED_CODEENGINE_CATALOG: OnceLock<SharedCodeEngineCatalogRecord> = OnceLock::new();
    SHARED_CODEENGINE_CATALOG.get_or_init(|| {
        serde_json::from_str(include_str!("../generated/engine-catalog.json"))
            .expect("parse generated codeengine catalog")
    })
}

pub fn list_codeengine_descriptors() -> Vec<CodeEngineDescriptorRecord> {
    shared_codeengine_catalog().engines.clone()
}

pub fn list_codeengine_model_catalog_entries() -> Vec<CodeEngineModelCatalogEntryRecord> {
    shared_codeengine_catalog().models.clone()
}

pub fn list_native_session_provider_catalog_entries() -> Vec<NativeSessionProviderCatalogRecord> {
    native_session_provider_catalog_entries().to_vec()
}

pub fn find_codeengine_descriptor(engine_key: &str) -> Option<CodeEngineDescriptorRecord> {
    let normalized_engine_key = engine_key.trim().to_ascii_lowercase();
    shared_codeengine_catalog()
        .engines
        .iter()
        .find(|engine| {
            engine
                .engine_key
                .eq_ignore_ascii_case(normalized_engine_key.as_str())
        })
        .cloned()
}

pub fn find_native_session_provider_catalog_entry(
    engine_id: &str,
) -> Option<&'static NativeSessionProviderCatalogRecord> {
    let normalized_engine_id = engine_id.trim().to_ascii_lowercase();
    native_session_provider_catalog_entries()
        .iter()
        .find(|provider| provider.engine_id.eq_ignore_ascii_case(normalized_engine_id.as_str()))
}

File: apps/sdkwork-birdcoder-pc/packages/sdkwork-birdcoder-pc-codeengine/src/catalog.ts
import type {
  BirdCoderCodeEngineKey,
  BirdCoderEngineCapabilityMatrix,
  BirdCoderEngineDescriptor,
  BirdCoderEngineOfficialEntry,
  BirdCoderEngineTransportKind,
  BirdCoderModelCatalogEntry,
} from '@sdkwork/birdcoder-pc-contracts-commons';

export const WORKBENCH_CODE_ENGINE_IDS = [
  'codex',
  'claude-code',
  'gemini',
  'opencode',
] as const;

export type WorkbenchCodeEngineId = (typeof WORKBENCH_CODE_ENGINE_IDS)[number];

export interface BirdCoderCodeEngineNativeSessionProvider {
  engineKey: WorkbenchCodeEngineId;
  engineId: WorkbenchCodeEngineId;
  displayName: string;
  prefix: string;
  nativeSessionIdPrefix: string;
  transportKinds: readonly BirdCoderEngineTransportKind[];
  discoveryMode: 'explicit-only' | 'passive-global';
  rawIdKind: string;
  description: string;
}

const CATALOG_TIMESTAMP = '2026-04-28T00:00:00.000Z';
const DEFAULT_TENANT_ID = '0';

const STANDARD_CAPABILITY_MATRIX: BirdCoderEngineCapabilityMatrix = {
  chat: true,
  streaming: true,
  structuredOutput: true,
  toolCalls: true,
  planning: true,
  patchArtifacts: true,
  commandArtifacts: true,
  todoArtifacts: true,
  ptyArtifacts: true,
  previewArtifacts: true,
  testArtifacts: true,
  approvalCheckpoints: true,
  sessionResume: true,
  remoteBridge: false,
  mcp: true,
};

function capabilityMatrix(
  overrides: Partial<BirdCoderEngineCapabilityMatrix> = {},
): BirdCoderEngineCapabilityMatrix {
  return {
    ...STANDARD_CAPABILITY_MATRIX,
    ...overrides,
  };
}

function officialEntry(input: BirdCoderEngineOfficialEntry): BirdCoderEngineOfficialEntry {
  return input;
}

function entitySummary(id: string): Pick<
  BirdCoderEngineDescriptor,
  'id' | 'uuid' | 'tenantId' | 'createdAt' | 'updatedAt'
> {
  return {
    id,
    uuid: `00000000-0000-4000-8000-${id.padStart(12, '0').slice(0, 12)}`,
    tenantId: DEFAULT_TENANT_ID,
    createdAt: CATALOG_TIMESTAMP,
    updatedAt: CATALOG_TIMESTAMP,
  };
}

const CODEX_CAPABILITY_MATRIX = capabilityMatrix({
  remoteBridge: false,
});

const CLAUDE_CAPABILITY_MATRIX = capabilityMatrix({
  remoteBridge: false,
});

const GEMINI_CAPABILITY_MATRIX = capabilityMatrix({
  remoteBridge: false,
  ptyArtifacts: false,
});

const OPENCODE_CAPABILITY_MATRIX = capabilityMatrix({
  structuredOutput: false,
  remoteBridge: false,
});

export const BIRDCODER_CODE_ENGINE_DESCRIPTORS: readonly BirdCoderEngineDescriptor[] = [
  {
    ...entitySummary('1000000000000000001'),
    engineKey: 'codex',
    displayName: 'Codex',
    vendor: 'OpenAI',
    installationKind: 'external-cli',
    defaultModelId: 'gpt-5.4',
    homepage: 'https://openai.com/codex',
    supportedHostModes: ['web', 'desktop', 'server'],
    transportKinds: ['cli-jsonl', 'sdk-stream'],
    capabilityMatrix: CODEX_CAPABILITY_MATRIX,
    status: 'active',
    accessPlan: {
      primaryLaneId: 'codex-cli-jsonl',
      fallbackLaneIds: ['codex-sdk-stream'],
      lanes: [
        {
          laneId: 'codex-sdk-stream',
          label: 'Codex official SDK',
          strategyKind: 'rust-native',
          runtimeOwner: 'rust-server',
          bridgeProtocol: 'direct',
          transportKind: 'sdk-stream',
          status: 'planned',
          enabledByDefault: false,
          hostModes: ['desktop', 'server'],
          description: 'OpenAI Codex official SDK lane; runtime promotion is pending live SDK conformance.',
        },
        {
          laneId: 'codex-cli-jsonl',
          label: 'Codex CLI JSONL',
          strategyKind: 'cli-spawn',
          runtimeOwner: 'rust-server',
          bridgeProtocol: 'stdio',
          transportKind: 'cli-jsonl',
          status: 'ready',
          enabledByDefault: true,
          hostModes: ['desktop', 'server'],
          description: 'OpenAI Codex CLI JSONL execution lane used by the local kernel runtime.',
        },
      ],
    },
    officialIntegration: {
      integrationClass: 'official-protocol',
      runtimeMode: 'headless',
      officialEntry: officialEntry({
        packageName: '@openai/codex-sdk',
        sdkPath: 'external/codex/sdk/typescript',
        cliPackageName: '@openai/codex',
        sourceMirrorPath: 'external/codex',
      }),
    },
  },
  {
    ...entitySummary('1000000000000000002'),
    engineKey: 'claude-code',
    displayName: 'Claude Code',
    vendor: 'Anthropic',
    installationKind: 'external-cli',
    defaultModelId: 'claude-sonnet-4-6',
    homepage: 'https://www.anthropic.com/claude-code',
    supportedHostModes: ['web', 'desktop', 'server'],
    transportKinds: ['cli-jsonl', 'sdk-stream', 'remote-control-http'],
    capabilityMatrix: CLAUDE_CAPABILITY_MATRIX,
    status: 'active',
    accessPlan: {
      primaryLaneId: 'claude-code-cli-print',
      fallbackLaneIds: ['claude-code-sdk-stream', 'claude-code-remote-control'],
      lanes: [
        {
          laneId: 'claude-code-sdk-stream',
          label: 'Claude Agent SDK',
          strategyKind: 'grpc-bridge',
          runtimeOwner: 'typescript-bridge',
          bridgeProtocol: 'grpc',
          transportKind: 'sdk-stream',
          status: 'planned',
          enabledByDefault: false,
          hostModes: ['desktop', 'server'],
          description: 'Anthropic Claude Agent SDK lane; runtime promotion is pending live SDK conformance.',
        },
        {
          laneId: 'claude-code-cli-print',
          label: 'Claude Code CLI print',
          strategyKind: 'cli-spawn',
          runtimeOwner: 'typescript-bridge',
          bridgeProtocol: 'stdio',
          transportKind: 'cli-jsonl',
          status: 'ready',
          enabledByDefault: true,
          hostModes: ['desktop', 'server'],
          description: 'Claude Code CLI print execution lane used by the local kernel runtime.',
        },
        {
          laneId: 'claude-code-remote-control',
          label: 'Claude Code remote control',
          strategyKind: 'remote-control',
          runtimeOwner: 'typescript-bridge',
          bridgeProtocol: 'http',
          transportKind: 'remote-control-http',
          status: 'planned',
          enabledByDefault: false,
          hostModes: ['web', 'desktop', 'server'],
          description: 'Claude Code remote-control lane; cloud transport is not implemented yet.',
        },
      ],
    },
    officialIntegration: {
      integrationClass: 'official-protocol',
      runtimeMode: 'headless',
      officialEntry: officialEntry({
        packageName: '@anthropic-ai/claude-agent-sdk',
        sdkPath: null,
        cliPackageName: 'claude-code',
        sourceMirrorPath: 'external/claude-code',
        supplementalLanes: ['claude-code-cli-print', 'claude-code-remote-control'],
      }),
    },
  },
  {
    ...entitySummary('1000000000000000003'),
    engineKey: 'gemini',
    displayName: 'Gemini',
    vendor: 'Google',
    installationKind: 'external-cli',
    defaultModelId: 'auto-gemini-3',
    homepage: 'https://github.com/google-gemini/gemini-cli',
    supportedHostModes: ['web', 'desktop', 'server'],
    transportKinds: ['cli-jsonl', 'sdk-stream'],
    capabilityMatrix: GEMINI_CAPABILITY_MATRIX,
    status: 'active',
    accessPlan: {
      primaryLaneId: 'gemini-cli-jsonl',
      fallbackLaneIds: ['gemini-sdk-stream'],
      lanes: [
        {
          laneId: 'gemini-sdk-stream',
          label: 'Gemini CLI SDK',
          strategyKind: 'grpc-bridge',
          runtimeOwner: 'typescript-bridge',
          bridgeProtocol: 'grpc',
          transportKind: 'sdk-stream',
          status: 'planned',
          enabledByDefault: false,
          hostModes: ['desktop', 'server'],
          description: 'Google Gemini CLI SDK lane; runtime promotion is pending live SDK conformance.',
        },
        {
          laneId: 'gemini-cli-jsonl',
          label: 'Gemini CLI JSONL',
          strategyKind: 'cli-spawn',
          runtimeOwner: 'typescript-bridge',
          bridgeProtocol: 'stdio',
          transportKind: 'cli-jsonl',
          status: 'ready',
          enabledByDefault: true,
          hostModes: ['desktop', 'server'],
          description: 'Gemini CLI JSONL execution lane used by the local kernel runtime.',
        },
      ],
    },
    officialIntegration: {
      integrationClass: 'official-protocol',
      runtimeMode: 'headless',
      officialEntry: officialEntry({
        packageName: '@google/gemini-cli-sdk',
        sdkPath: 'external/gemini/packages/sdk',
        cliPackageName: '@google/gemini-cli',
        sourceMirrorPath: 'external/gemini',
      }),
    },
  },
  {
    ...entitySummary('1000000000000000004'),
    engineKey: 'opencode',
    displayName: 'OpenCode',
    vendor: 'OpenCode',
    installationKind: 'external-cli',
    defaultModelId: 'opencode/big-pickle',
    homepage: 'https://opencode.ai',
    supportedHostModes: ['web', 'desktop', 'server'],
    transportKinds: ['cli-jsonl', 'sdk-stream', 'openapi-http'],
    capabilityMatrix: OPENCODE_CAPABILITY_MATRIX,
    status: 'active',
    accessPlan: {
      primaryLaneId: 'opencode-cli-jsonl',
      fallbackLaneIds: ['opencode-sdk-stream', 'opencode-openapi-http'],
      lanes: [
        {
          laneId: 'opencode-sdk-stream',
          label: 'OpenCode SDK',
          strategyKind: 'grpc-bridge',
          runtimeOwner: 'rust-server',
          bridgeProtocol: 'grpc',
          transportKind: 'sdk-stream',
          status: 'planned',
          enabledByDefault: false,
          hostModes: ['desktop', 'server'],
          description: 'OpenCode SDK lane; runtime promotion is pending live SDK conformance.',
        },
        {
          laneId: 'opencode-openapi-http',
          label: 'OpenCode OpenAPI HTTP',
          strategyKind: 'openapi-proxy',
          runtimeOwner: 'rust-server',
          bridgeProtocol: 'http',
          transportKind: 'openapi-http',
          status: 'planned',
          enabledByDefault: false,
          hostModes: ['web', 'desktop', 'server'],
          description: 'OpenCode OpenAPI lane; remote HTTP transport is not implemented yet.',
        },
        {
          laneId: 'opencode-cli-jsonl',
          label: 'OpenCode CLI JSONL',
          strategyKind: 'cli-spawn',
          runtimeOwner: 'rust-server',
          bridgeProtocol: 'stdio',
          transportKind: 'cli-jsonl',
          status: 'ready',
          enabledByDefault: true,
          hostModes: ['desktop', 'server'],
          description: 'OpenCode CLI JSON execution lane used by the local kernel runtime.',
        },
      ],
    },
    officialIntegration: {
      integrationClass: 'official-protocol',
      runtimeMode: 'headless',
      officialEntry: officialEntry({
        packageName: '@opencode-ai/sdk',
        sdkPath: 'external/opencode/packages/sdk/js',
        cliPackageName: 'opencode-ai',
        sourceMirrorPath: 'external/opencode',
      }),
    },
  },
];

const MODEL_CATALOG_INPUTS: ReadonlyArray<{
  engineKey: WorkbenchCodeEngineId;
  modelId: string;
  displayName: string;
  providerId?: string;
  defaultForEngine?: boolean;
  capabilityMatrix?: Partial<BirdCoderEngineCapabilityMatrix>;
}> = [
  {
    engineKey: 'codex',
    modelId: 'gpt-5.4',
    displayName: 'GPT-5.4',
    providerId: 'openai',
    defaultForEngine: true,
  },
  {
    engineKey: 'codex',
    modelId: 'gpt-5.3-codex',
    displayName: 'GPT-5.3 Codex',
    providerId: 'openai',
  },
  {
    engineKey: 'codex',
    modelId: 'gpt-5.2-codex',
    displayName: 'GPT-5.2 Codex',
    providerId: 'openai',
  },
  {
    engineKey: 'codex',
    modelId: 'gpt-5.1-codex',
    displayName: 'GPT-5.1 Codex',
    providerId: 'openai',
  },
  {
    engineKey: 'codex',
    modelId: 'gpt-5.5',
    displayName: 'GPT-5.5',
    providerId: 'openai',
  },
  {
    engineKey: 'claude-code',
    modelId: 'claude-sonnet-4-6',
    displayName: 'Claude Sonnet 4.6',
    providerId: 'anthropic',
    defaultForEngine: true,
  },
  {
    engineKey: 'claude-code',
    modelId: 'claude-code',
    displayName: 'Claude Code',
    providerId: 'anthropic',
  },
  {
    engineKey: 'claude-code',
    modelId: 'claude-3-opus',
    displayName: 'Claude 3 Opus',
    providerId: 'anthropic',
  },
  {
    engineKey: 'gemini',
    modelId: 'auto-gemini-3',
    displayName: 'Auto Gemini 3',
    providerId: 'google',
    defaultForEngine: true,
  },
  {
    engineKey: 'gemini',
    modelId: 'auto-gemini-2.5',
    displayName: 'Auto Gemini 2.5',
    providerId: 'google',
  },
  {
    engineKey: 'gemini',
    modelId: 'gemini-3.1-pro-preview',
    displayName: 'Gemini 3.1 Pro Preview',
    providerId: 'google',
  },
  {
    engineKey: 'gemini',
    modelId: 'gemini-3.1-pro-preview-customtools',
    displayName: 'Gemini 3.1 Pro Preview Custom Tools',
    providerId: 'google',
  },
  {
    engineKey: 'gemini',
    modelId: 'gemini-3.1-flash-lite-preview',
    displayName: 'Gemini 3.1 Flash Lite Preview',
    providerId: 'google',
  },
  {
    engineKey: 'gemini',
    modelId: 'gemini-3-pro-preview',
    displayName: 'Gemini 3 Pro Preview',
    providerId: 'google',
  },
  {
    engineKey: 'gemini',
    modelId: 'gemini-3-flash-preview',
    displayName: 'Gemini 3 Flash Preview',
    providerId: 'google',
  },
  {
    engineKey: 'gemini',
    modelId: 'gemini-2.5-pro',
    displayName: 'Gemini 2.5 Pro',
    providerId: 'google',
  },
  {
    engineKey: 'gemini',
    modelId: 'gemini-2.5-flash',
    displayName: 'Gemini 2.5 Flash',
    providerId: 'google',
  },
  {
    engineKey: 'gemini',
    modelId: 'gemini-2.5-flash-lite',
    displayName: 'Gemini 2.5 Flash Lite',
    providerId: 'google',
  },
  {
    engineKey: 'opencode',
    modelId: 'opencode/big-pickle',
    displayName: 'OpenCode Big Pickle',
    providerId: 'opencode',
    defaultForEngine: true,
  },
  {
    engineKey: 'opencode',
    modelId: 'opencode',
    displayName: 'OpenCode',
    providerId: 'opencode',
  },
  {
    engineKey: 'opencode',
    modelId: 'opencode/claude-sonnet-4-6',
    displayName: 'OpenCode Claude Sonnet 4.6',
    providerId: 'opencode',
  },
];

export const BIRDCODER_CODE_ENGINE_MODELS: readonly BirdCoderModelCatalogEntry[] =
  MODEL_CATALOG_INPUTS.map((input, index) => {
    const descriptor = BIRDCODER_CODE_ENGINE_DESCRIPTORS.find(
      (engine) => engine.engineKey === input.engineKey,
    );
    const transportKinds: readonly BirdCoderEngineTransportKind[] =
      descriptor?.transportKinds ?? ['sdk-stream'];

    return {
      ...entitySummary(String(2000000000000000000n + BigInt(index + 1))),
      engineKey: input.engineKey,
      modelId: input.modelId,
      displayName: input.displayName,
      providerId: input.providerId,
      status: 'active',
      defaultForEngine: Boolean(input.defaultForEngine),
      transportKinds,
      capabilityMatrix: input.capabilityMatrix ?? {},
    };
  });

export const BIRDCODER_CODE_ENGINE_NATIVE_SESSION_PROVIDERS: readonly BirdCoderCodeEngineNativeSessionProvider[] =
  [
    {
      engineKey: 'codex',
      engineId: 'codex',
      displayName: 'Codex',
      prefix: 'codex-native:',
      nativeSessionIdPrefix: 'codex-native:',
      transportKinds: ['cli-jsonl', 'sdk-stream'],
      discoveryMode: 'passive-global',
      rawIdKind: 'uuid',
      description: 'Codex provider-native session id.',
    },
    {
      engineKey: 'claude-code',
      engineId: 'claude-code',
      displayName: 'Claude Code',
      prefix: 'claude-code-native:',
      nativeSessionIdPrefix: 'claude-code-native:',
      transportKinds: ['cli-jsonl', 'sdk-stream', 'remote-control-http'],
      discoveryMode: 'passive-global',
      rawIdKind: 'string',
      description: 'Claude Code provider-native session id.',
    },
    {
      engineKey: 'gemini',
      engineId: 'gemini',
      displayName: 'Gemini',
      prefix: 'gemini-native:',
      nativeSessionIdPrefix: 'gemini-native:',
      transportKinds: ['cli-jsonl', 'sdk-stream'],
      discoveryMode: 'passive-global',
      rawIdKind: 'string',
      description: 'Gemini provider-native session id.',
    },
    {
      engineKey: 'opencode',
      engineId: 'opencode',
      displayName: 'OpenCode',
      prefix: 'opencode-native:',
      nativeSessionIdPrefix: 'opencode-native:',
      transportKinds: ['cli-jsonl', 'sdk-stream', 'openapi-http'],
      discoveryMode: 'passive-global',
      rawIdKind: 'string',
      description: 'OpenCode provider-native session id.',
    },
  ];

export function listBirdCoderCodeEngineDescriptors(): readonly BirdCoderEngineDescriptor[] {
  return BIRDCODER_CODE_ENGINE_DESCRIPTORS;
}

export function listBirdCoderCodeEngineModels(): readonly BirdCoderModelCatalogEntry[] {
  return BIRDCODER_CODE_ENGINE_MODELS;
}

export function listBirdCoderCodeEngineNativeSessionProviders():
  readonly BirdCoderCodeEngineNativeSessionProvider[] {
  return BIRDCODER_CODE_ENGINE_NATIVE_SESSION_PROVIDERS;
}

export function getBirdCoderCodeEngineDescriptor(
  engineKey: BirdCoderCodeEngineKey | null | undefined,
): BirdCoderEngineDescriptor | null {
  const key = String(engineKey ?? '').trim();

  return (
    BIRDCODER_CODE_ENGINE_DESCRIPTORS.find((descriptor) => descriptor.engineKey === key) ??
    null
  );
}

export function getBirdCoderCodeEngineCapabilities(
  engineKey: BirdCoderCodeEngineKey | null | undefined,
): BirdCoderEngineCapabilityMatrix | null {
  return getBirdCoderCodeEngineDescriptor(engineKey)?.capabilityMatrix ?? null;
}

export function resolveBirdCoderCodeEngineNativeSessionIdPrefix(
  engineKey: BirdCoderCodeEngineKey | null | undefined,
): string | null {
  const key = String(engineKey ?? '').trim();

  return (
    BIRDCODER_CODE_ENGINE_NATIVE_SESSION_PROVIDERS.find(
      (provider) => provider.engineKey === key,
    )?.prefix ?? null
  );
}

export function normalizeBirdCoderCodeEngineNativeSessionId(
  nativeSessionId: unknown,
  _engineKey?: BirdCoderCodeEngineKey | null,
): string {
  let value = String(nativeSessionId ?? '').trim();

  for (const provider of BIRDCODER_CODE_ENGINE_NATIVE_SESSION_PROVIDERS) {
    if (value.startsWith(provider.prefix)) {
      value = value.slice(provider.prefix.length);
      break;
    }
  }

  return value.trim();
}

export function isBirdCoderCodeEngineNativeSessionId(
  value: unknown,
  _engineKey?: BirdCoderCodeEngineKey | null,
): boolean {
  const normalizedValue = String(value ?? '').trim();
  return BIRDCODER_CODE_ENGINE_NATIVE_SESSION_PROVIDERS.some(
    (provider) =>
      normalizedValue.startsWith(provider.prefix) &&
      normalizedValue.slice(provider.prefix.length).trim().length > 0,
  );
}
